Client/WebFilters/CosmeticFiltersResourceDownloader.swift

```diff
@@ -64,109 +66,122 @@ class CosmeticFiltersResourceDownloader {
+            log.error("Failed to Download Cosmetic-Filters: \(error)")
+          default:
+            break
           }
-        } catch {
-          log.error("Failed to Download Cosmetic-Filters: \(error)")
+        } receiveValue: { [weak self] engine in
+          log.debug("Successfully Downloaded Cosmetic-Filters")
+          self?.engine = engine
         }
-      }
     }
   }
 
   func cssRules(for url: URL) -> String? {
     engine.cssRules(for: url)
   }
 
-  private func loadDownloadedFiles(into engine: AdblockRustEngine) async throws {
+  private func loadDownloadedFiles(into engine: AdblockRustEngine) -> AnyPublisher<Void, Error> {
     let fm = FileManager.default
     guard let folderUrl = fm.getOrCreateFolder(name: CosmeticFiltersResourceDownloader.folderName) else {
-      log.error("Could not get directory with .dat and .json files")
-      return
+      return Fail(error: "Could not get directory with .dat and .json files").eraseToAnyPublisher()
     }
 
     let enumerator = fm.enumerator(at: folderUrl, includingPropertiesForKeys: nil)
     let filePaths = enumerator?.allObjects as? [URL]
     let datFileUrls = filePaths?.filter { $0.pathExtension == "dat" }
     let jsonFileUrls = filePaths?.filter { $0.pathExtension == "json" }
 
-    try await datFileUrls?.asyncForEach({
+    let dataFilesSetup: [AnyPublisher<Void, Error>] = datFileUrls?.compactMap({
       let fileName = $0.deletingPathExtension().lastPathComponent
-      if let data = fm.contents(atPath: $0.path) {
-        try await self.setDataFile(into: engine, data: data, id: fileName)
-      }
-    })
+      guard let data = fm.contents(atPath: $0.path) else { return nil }
+      return self.setDataFile(into: engine, data: data, id: fileName)
+    }) ?? []
 
-    try await jsonFileUrls?.asyncForEach {
+    let jsonFilesSetup: [AnyPublisher<Void, Error>] = jsonFileUrls?.compactMap({
       let fileName = $0.deletingPathExtension().lastPathComponent
-      if let data = fm.contents(atPath: $0.path) {
-        try await self.setJSONFile(into: engine, data: data, id: fileName)
-      }
-    }
+      guard let data = fm.contents(atPath: $0.path) else { return nil }
+      return self.setJSONFile(into: engine, data: data, id: fileName)
+    }) ?? []
+    
+    return Publishers.MergeMany(dataFilesSetup + jsonFilesSetup)
+      .collect()
+      .subscribe(on: DispatchQueue.global(qos: .userInitiated))
+      .flatMap { $0.publisher }
+      .eraseToAnyPublisher()
   }
 
-  private func downloadCosmeticSamples(with engine: AdblockRustEngine) async throws {
-    try await downloadResources(for: engine, type: .cosmeticSample)
-    log.debug("Downloaded Cosmetic Filters CSS Samples")
-    Preferences.Debug.lastCosmeticFiltersCSSUpdate.value = Date()
+  private func downloadCosmeticSamples(with engine: AdblockRustEngine) -> AnyPublisher<Void, Error> {
+    downloadResources(for: engine, type: .cosmeticSample)
+      .receive(on: DispatchQueue.main)
+      .map {
```

Comment:
what does it `map` to? shouldn't sink be used here or something else

Reply:
It maps to nothing. The `downloadResources` function just returns `Void` already. So when I called `map` here, it's just like calling `.then` on a `promise`. So no need for sink here because I'm not returning a `AnyCancellable`. This function is executed as part of the chain of completion :)